addon/-private/system/identity-map.js

```diff
@@ -0,0 +1,49 @@
+import RecordMap from './record-map';
+
+/**
+ `IdentityMap` is a custom storage map for records by modelName
+ used by `DS.Store`.
+
+ @class IdentityMap
+ @private
+ */
+export default class IdentityMap {
+  constructor() {
+    this._map = Object.create(null);
+  }
+
+  /**
+   Retrieves the `RecordMap` for a given modelName,
+   creating one if one did not already exist. This is
+   similar to `getWithDefault` or `get` on a `MapWithDefault`
+
+   @method retrieve
+   @param modelName a previously normalized modelName
+   @returns {RecordMap} the RecordMap for the given modelName
+   */
+  retrieve(modelName) {
+    let recordMap = this._map[modelName];
+
+    if (!recordMap) {
+      recordMap = this._map[modelName] = new RecordMap(modelName);
+    }
+
+    return recordMap;
+  }
+
+  /**
+   Clears the contents of all known `RecordMaps`, but does
+   not remove the RecordMap instances.
+
+   @method clear
+   */
+  clear() {
+    let recordMaps = this._map;
```

Comment:
I believe we must specifically clear them because clear is responsible for tearing down the internal models.  This is a refactor that preserves existing logic for teardown, but it is absolutely possible we were doing an excessive amount of work on teardown in the past.